Analysis/Tasks/PWGHF/HFCandidateCreatorX.cxx

```diff
@@ -0,0 +1,300 @@
+                           std::sqrt(impactParameter0.getSigmaY2()), std::sqrt(impactParameter1.getSigmaY2()), std::sqrt(impactParameter2.getSigmaY2()),
+                           jpsiCand.globalIndex(), trackPos.globalIndex(), trackNeg.globalIndex(),
+                           hfFlag);
+
+          // calculate invariant mass
+          auto arrayMomenta = array{pvecJpsi, pvecPos, pvecNeg};
+          massJpsiPiPi = RecoDecay::M(std::move(arrayMomenta), array{massJpsi, massPi, massPi});
+          hMassX->Fill(massJpsiPiPi);
+        } // pi- loop
+      }   // pi+ loop
+    }     // Jpsi loop
+  }       // process
+};        // struct
+
+/// Extends the base table with expression columns.
+struct HFCandidateCreatorXExpressions {
+  Spawns<aod::HfCandXExt> rowCandidateX;
+  void init(InitContext const&) {}
+};
+
+/// Performs MC matching.
+struct HFCandidateCreatorXMC {
+  Produces<aod::HfCandXMCRec> rowMCMatchRec;
+  Produces<aod::HfCandXMCGen> rowMCMatchGen;
+
+  void process(aod::HfCandX const& candidates,
+               aod::HfCandProng2,
+               aod::BigTracksMC const& tracks,
+               aod::McParticles const& particlesMC)
+  {
+    int indexRec = -1;
+    int8_t sign = 0;
+    int8_t flag = 0;
+    int8_t origin = 0;
+    int8_t channel = 0;
+
+    // Match reconstructed candidates.
+    for (auto& candidate : candidates) {
+      //Printf("New rec. candidate");
+      flag = 0;
+      origin = 0;
+      channel = 0;
+      auto jpsiTrack = candidate.index0();
+      auto arrayDaughters = array{candidate.index1_as<aod::BigTracksMC>(),
+                                  candidate.index2_as<aod::BigTracksMC>(),
+                                  jpsiTrack.index0_as<aod::BigTracksMC>(),
+                                  jpsiTrack.index1_as<aod::BigTracksMC>()};
+
+      // X → J/ψ π+ π-
+      //Printf("Checking X → J/ψ π+ π-");
+      indexRec = RecoDecay::getMatchedMCRec(particlesMC, arrayDaughters, 9920443, array{+kPiPlus, -kPiPlus, +kElectron, -kElectron}, true, &sign, 2);
+      if (indexRec > -1) {
+        flag = 1 << XToJpsiPiPi;
+      }
+
+      // Check whether the particle is non-prompt (from a b quark).
+      if (flag != 0) {
+        auto particle = particlesMC.iteratorAt(indexRec);
+        origin = (RecoDecay::getMother(particlesMC, particle, 5, true) > -1 ? NonPrompt : Prompt);
+      }
+
+      rowMCMatchRec(flag, origin, channel);
+    }
+
+    // Match generated particles.
+    for (auto& particle : particlesMC) {
+      //Printf("New gen. candidate");
+      flag = 0;
+      origin = 0;
+      channel = 0;
+
+      // X → J/ψ π+ π-
+      //Printf("Checking X → J/ψ π+ π-");
+      if (RecoDecay::isMatchedMCGen(particlesMC, particle, 9920443, array{443, +kPiPlus, -kPiPlus}, true)) {
+        // Match J/psi --> e+e-
+        std::vector<int> arrDaughter;
+        RecoDecay::getDaughters(particlesMC, particle, &arrDaughter, array{443}, 1);
+        auto jpsiCandMC = particlesMC.iteratorAt(arrDaughter[0]);
```

Comment:
You can get the daughters from `isMatchedMCGen` directly.
How do you know that J/ψ is on index 0?